### ISSUE TITLE
"glide guess" is broken with Go 1.5 official package installer

### ISSUE THREAD

**Issue by janeczku** [Masterminds/glide#74]
### Steps to reproduce

1. Remove previous Go version and any `$GOROOT` declarations in .bash_profile/.zshrc
2. Install go1.5.darwin-amd64.pkg from golang.org
3. Switch to any project folder within the `$GOPATH`
4. Run `glide create`
5. Run `glide guess`

### Expected result
`glide guess` does not fail

### Actual result
`glide guess` fails with error:

```sh
Oops! cannot find package "fmt" in any of:
	/usr/local/Cellar/go/1.5/libexec/src/fmt (from $GOROOT)
	/Users/Janek/go/src/fmt (from $GOPATH)
```

### Additional information

`/usr/local/Cellar` does not exist on my system. It seems to be the path were Go would be installed using `brew`.

`$GOROOT` env variable is **not** set on my system. This is not necessary with the package installer, since it installs Go to the default location used by the Go binary.

The Go binary correctly uses this location as its root path:

```sh
$ go env
(...)
GOPATH="/Users/Janek/go"
GOROOT="/usr/local/go"
```

It looks like `glide` always falls back to the path where `brew` installs Go. Obviously, when Go is not installed using `brew` this will break things.

**Proposed solution:**
Parse GOROOT from `go env` output.

**Comment by mattfarina** [comment]
Thanks for such a great bug report!

The issue is in the `go/build` package and how it looks for packages. We'll need to deal with the situation where no `$GOROOT` is set. The go toolchain has diverged from this package.

I'll fix this in the next day or two.

**Comment by janeczku** [comment]
@mattfarina  see above, even with $GOROOT set i still get the error.

**Comment by janeczku** [comment]
@mattfarina  Ups my error. Setting GOROOT solves the problem!

**Comment by mattfarina** [comment]
This is a bit complicated and I need a little feedback.

We discover the imports using the `go/build` package. This package knows the environment including the `$GOROOT` which it picks up from the environment. If your `$GOROOT` isn't set there will be errors.

So, I'm confused how you got this error with the default installer. I tried it with both the 1.5 and 1.5.1 installers and was unable to reproduce the problem. When I intentionally misconfigured my `$GOROOT` the `go` tool gave me errors.

Can you try this again with 1.5.1 which was just released? There was issue with the `go/build` package relating to the `$GOROOT` that was fixed in this release.

**Comment by technosophos** [comment]
@janeczku If you have a moment or two for some experimentation, I've been looking into building more tooling around figuring out what dependencies are used. Today I committed `glide tree` and `glide list` as commands that inspect code, like `glide guess`.

- Tree presents package imports organized as a tree.
- List lists all of the packages that a project imports.

If you're up for giving them a test run, I'd really appreciate it.

**Comment by technosophos** [comment]
I'm also chatting with @mattfarina about whether we ought to add the capability to `glide guess` that would cause it to automatically fetch (into vendor/) and packages that are not already present. Something like `glide guess --get`. Is that something that you'd find helpful? Any thoughts on that?

**Comment by janeczku** [comment]
@mattfarina 
> We discover the imports using the go/build package. This package knows the environment including the $GOROOT which it picks up from the environment. If your $GOROOT isn't set there will be errors.

Again, with the Go package installer $GOROOT does not need to be set. Here is why:

From http://golang.org/pkg/runtime/

```
func GOROOT() string
GOROOT returns the root of the Go tree. It uses the GOROOT environment variable,
if set, or else the root used during the Go build.
```

That indicates that the Go runtime falls back to the root path were it was build in when the env var is not set. The packages installer installs Go in that same path. Thats why GOROOT env var is not needed.

By the way, without GOROOT env var both the go/build and runtime packages returns the correct root path on my system:

```go
package main

import (
	"fmt"
	"go/build"
	"runtime"
)

var buildContext = build.Default

func main() {
	fmt.Println("GOROOT runtime: ", runtime.GOROOT())
	fmt.Println("GOROOT go/build: ", buildContext.GOROOT)
}
```

Result:
```
GOROOT Runtime:  /usr/local/go
GOROOT Build Context:  /usr/local/go
```

**Comment by janeczku** [comment]
@technosophos 
Having `glide guess --get` would definitely be useful to me - especially when converting an existing project to glide.
I will test drive the new commands later this day.

**Comment by mattfarina** [comment]
@janeczku I've installed Go from the installer on golang.org and tried glide guess without an issue. I'm having trouble reproducing the problem.

I'm using the tip of master. Which version are you using? Is there something else I should be looking at?

Dependencies are basically [imported using `go/build`](https://github.com/Masterminds/glide/blob/f21a3de2393c0013d01faeb4ed5d8c43fd9205f4/cmd/guess_deps.go#L54) which knows the right location for the `GOROOT`.

I'm having trouble tracking down your problem. Any additional help would be immensely appreciated.

**Comment by janeczku** [comment]
@mattfarina your $GOROOT env var was unset when you tested this?

**Comment by janeczku** [comment]
@mattfarina
```sh
Oops! cannot find package "fmt" in any of:
	/usr/local/Cellar/go/1.5/libexec/src/fmt (from $GOROOT)
	/Users/Janek/go/src/fmt (from $GOPATH)
```
I don't even have `brew` installed on my Mac. Where is that `/usr/local/Cellar` coming from? Definitely not  from $GOROOT since that env var is unset.

**Comment by mattfarina** [comment]
@janeczku I finally figured it out. The way to test this is to use glide built from a homebrew installed Go on system where Go was installed via the installer on golang.org. This is specific to the binary distribution method. The binary appears to have the wrong default `$GOPATH` when one is not set.

I think the fix for this is going to require us to create our own import context rather than using the default one in the `go/build` package.

Thanks for being patient with me while I figured this out.
